### PR TITLE
Secure, configurable anonymisation

### LINKED ISSUE #79: Improve Anonymisation Process
Currently patient data is anonymised using SHA512 before being sent to the Inference service. There is potential risk of malicious actors reverse engineering this hashing and being able to link scans with specific hospitals and/or patients. Either salting needs to be added to the hashing, or the patient data should be removed in the same way as many other tags are. Further investigation is required on the pros/cons of each of theses options.

### PR DESCRIPTION
**Pull request by peterhessey**
Closes #79 and #104.

This PR introduces changes to how the anonymisation works in the IE-Gateway. Specifically, the DICOM tags sent to the inference service are now specified in `GatewayProcessorConfig.json` instead of `SegmenationClient.cs`, making it much simpler for users to configure their own anonymisation protocols. 

Furthermore, `PatientID` is no longer sent to the inference service by default, as it is not needed for inference.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -277,7 +278,8 @@ Note that the installer will include the configuration files in the folder [./So
 
 By default the installer will put the files in the folder: `C:\Program Files\Microsoft InnerEye Gateway`, with the following folder structure (with the default GatewayModelRulesConfig):
 
-```
+```shell
```
Comment (ant0nsc) [inline README.md:281]:
```suggestion
```text
```

File: Source/Microsoft.Gateway/Microsoft.InnerEye.Azure.Segmentation.Client/InnerEyeSegmentationClient.cs [hunk 2]
```diff
@@ -399,6 +327,51 @@ private static AnonymizeEngine GetAnonymisationEngine(Guid anonymisationProtocol
...
 
+        /// <inheritdoc />
+        public IEnumerable<DicomTagAnonymisation> GetSegmentationAnonymisationProtocol()
+        {
+            var parsedDicomTags = new List<DicomTagAnonymisation>();
+            foreach (var dicomTagConfig in _segmentationAnonymisationProtocol)
+            {
+                AnonymisationMethod anonMethod;
+
+                switch (dicomTagConfig["AnonymisationMethod"])
```
Comment (ant0nsc) [inline Source/Microsoft.Gateway/Microsoft.InnerEye.Azure.Segmentation.Client/InnerEyeSegmentationClient.cs:338]: there is no error handling for key not found

File: Source/Microsoft.Gateway/Microsoft.InnerEye.Azure.Segmentation.Client/InnerEyeSegmentationClient.cs [hunk 3]
```diff
@@ -399,6 +327,51 @@ private static AnonymizeEngine GetAnonymisationEngine(Guid anonymisationProtocol
...
+                {
+                    case "Keep":
```
Comment (ant0nsc) [inline Source/Microsoft.Gateway/Microsoft.InnerEye.Azure.Segmentation.Client/InnerEyeSegmentationClient.cs:340]: For all the 3 enum cases, should that be case insensitive?

File: Source/Microsoft.Gateway/Microsoft.InnerEye.Azure.Segmentation.Client/InnerEyeSegmentationClient.cs [hunk 4]
```diff
@@ -399,6 +327,51 @@ private static AnonymizeEngine GetAnonymisationEngine(Guid anonymisationProtocol
...
+                        anonMethod = AnonymisationMethod.RandomiseDateTime;
+                        break;
+                    default:
+                        throw new ArgumentException($"Invalid value for AnonymisationMethod found in Anonymisation config: {dicomTagConfig["AnonymisationMethod"]}. Permitted options are: \"Keep\", \"Hash\" and \"Random\"");
+                }
+
+                var fields = typeof(DicomTag).GetFields();
+                foreach (var field in fields)
+                {
+                    if (string.Equals(field.Name, dicomTagConfig["DicomTagID"], StringComparison.Ordinal))
```
Comment (ant0nsc) [inline Source/Microsoft.Gateway/Microsoft.InnerEye.Azure.Segmentation.Client/InnerEyeSegmentationClient.cs:356]: what if "DicomTagID" is not present?

File: Source/Microsoft.Gateway/Microsoft.InnerEye.Azure.Segmentation.Client/InnerEyeSegmentationClient.cs [hunk 5]
```diff
@@ -399,6 +327,51 @@ private static AnonymizeEngine GetAnonymisationEngine(Guid anonymisationProtocol
...
+                        parsedDicomTags.Add(new DicomTagAnonymisation(
+                          dicomTag: (DicomTag)field.GetValue(null),
+                          anonymisationProtocol: anonMethod)
+                        );
+                        break;
+                    }
+                }
+                if (parsedDicomTags.Count == 0)
+                {
+                    throw new ArgumentException($"Unknown DicomTag found in config: {dicomTagConfig["DicomTagID"]}");
```
Comment (ant0nsc) [inline Source/Microsoft.Gateway/Microsoft.InnerEye.Azure.Segmentation.Client/InnerEyeSegmentationClient.cs:367]: there are a lot of repeated string literals here

File: Source/Microsoft.Gateway/Microsoft.InnerEye.Gateway.Models/AnonymisationSettings.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+        public override bool Equals(object obj)
+        {
+            return Equals(obj as AnonymisationSettings);
+        }
+
+        /// <inheritdoc/>
+        public bool Equals(AnonymisationSettings other)
+        {
+            return other != null;  /// not sure about this! will return equal even if settings aren't the same?
```
Comment (ant0nsc) [inline Source/Microsoft.Gateway/Microsoft.InnerEye.Gateway.Models/AnonymisationSettings.cs:34]: I'm confused about the comment and the code

**Review by ant0nsc — CHANGES_REQUESTED** [review]
Looking good, overall - you really have picked up a lot of the C# bits!
I have not been able to spot a unit test for the core functionality of converting a string tuple (name, method) to its typed counterparts? (GetSegmentationAnonymisationProtocol) Anything to do with reflection should be tested really carefully, I feel. Also, this is a user-provided file, and should be checked more thoroughly for duplicates, etc.

At the risk of distracting you: Would an alternative way of creating  the JSON config lead to shorter and easier to maintain configs?
```
{
   "Keep": ["foo", "bar"],
   "randomize": ["date"],
   "hash": ["baz"]
}
```

File: Source/Microsoft.Gateway/Microsoft.InnerEye.Azure.Segmentation.Client/InnerEyeSegmentationClient.cs (hunk 5 above)
Comment (ant0nsc) [reply Source/Microsoft.Gateway/Microsoft.InnerEye.Azure.Segmentation.Client/InnerEyeSegmentationClient.cs:367]: turn them into constants?

**Comment by peterhessey** [comment]
> Looking good, overall - you really have picked up a lot of the C# bits! I have not been able to spot a unit test for the core functionality of converting a string tuple (name, method) to its typed counterparts? (GetSegmentationAnonymisationProtocol) Anything to do with reflection should be tested really carefully, I feel. Also, this is a user-provided file, and should be checked more thoroughly for duplicates, etc.
> 
> At the risk of distracting you: Would an alternative way of creating the JSON config lead to shorter and easier to maintain configs?
> 
> ```
> {
>    "Keep": ["foo", "bar"],
>    "randomize": ["date"],
>    "hash": ["baz"]
> }
> ```

I completely agree, and have updated the PR accordingly.

**Review by ant0nsc — APPROVED** [review]
Looking good! Just one question: What happens if someone has an existing Gateway installation, with an "old" config file, and they now use the new gateway code? Would it throw an error because there is no anonymization section?

**Review by javier-alvarez — APPROVED** [review]

File: Source/Microsoft.Gateway/Microsoft.InnerEye.Gateway.Models/AnonymisationSettings.cs [hunk 7]
```diff
@@ -11,34 +11,31 @@ public class AnonymisationSettings : IEquatable<AnonymisationSettings>
...
         {
             return Equals(obj as AnonymisationSettings);
         }
 
         /// <inheritdoc/>
-        public bool Equals(AnonymisationSettings other)
-        {
-            return other != null;  /// not sure about this! will return equal even if settings aren't the same?
-        }
+        public bool Equals(AnonymisationSettings other) => other != null && DicomTagsAnonymisationConfig.Count == other.DicomTagsAnonymisationConfig.Count;
```
[the remark this answers is not in the document]
Comment (peterhessey) [reply Source/Microsoft.Gateway/Microsoft.InnerEye.Gateway.Models/AnonymisationSettings.cs]: Good spot - had meant to come back to this as I got stuck on it. Fixed now, learning the hard way that `==` is really is not the same as in python!

**Pull request merged**